backend/groth16/bls12-377/mpcsetup/phase2.go

```diff
@@ -6,48 +6,200 @@
 package mpcsetup
 
 import (
+	"bytes"
 	"crypto/sha256"
 	"errors"
-	"math/big"
-
+	"fmt"
 	curve "github.com/consensys/gnark-crypto/ecc/bls12-377"
 	"github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/mpcsetup"
+	gcUtils "github.com/consensys/gnark-crypto/utils"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/backend/groth16/internal"
 	"github.com/consensys/gnark/constraint"
 	cs "github.com/consensys/gnark/constraint/bls12-377"
+	"github.com/consensys/gnark/internal/utils"
+	"math/big"
+	"slices"
+	"sync"
 )
 
-type Phase2Evaluations struct {
+// Phase2Evaluations components of the circuit keys
+// not depending on Phase2 randomisations
+type Phase2Evaluations struct { // TODO @Tabaie rename
 	G1 struct {
-		A, B, VKK []curve.G1Affine
+		A   []curve.G1Affine   // A are the left coefficient polynomials for each witness element, evaluated at τ
+		B   []curve.G1Affine   // B are the right coefficient polynomials for each witness element, evaluated at τ
+		VKK []curve.G1Affine   // VKK are the coefficients of the public witness and commitments
+		CKK [][]curve.G1Affine // CKK are the coefficients of the committed values
 	}
 	G2 struct {
-		B []curve.G2Affine
+		B []curve.G2Affine // B are the right coefficient polynomials for each witness element, evaluated at τ
 	}
+	PublicAndCommitmentCommitted [][]int
 }
 
 type Phase2 struct {
 	Parameters struct {
 		G1 struct {
-			Delta curve.G1Affine
-			L, Z  []curve.G1Affine
+			Delta    curve.G1Affine
+			Z        []curve.G1Affine   // Z[i] = xⁱt(x)/δ where t is the domain vanishing polynomial 0 ≤ i ≤ N-2
+			PKK      []curve.G1Affine   // PKK are the coefficients of the private witness, needed for the proving key. They have a denominator of δ
+			SigmaCKK [][]curve.G1Affine // Commitment proof bases: SigmaCKK[i][j] = σᵢCᵢⱼ where Cᵢⱼ is the commitment basis for the jᵗʰ committed element from the iᵗʰ commitment
 		}
 		G2 struct {
 			Delta curve.G2Affine
+			Sigma []curve.G2Affine // the secret σ value for each commitment
+		}
+	}
+
+	// Proofs of update correctness
+	Sigmas []mpcsetup.UpdateProof
+	Delta  mpcsetup.UpdateProof
+
+	// Challenge is the hash of the PREVIOUS contribution
+	Challenge []byte
+}
+
+func (p *Phase2) Verify(next *Phase2, options ...verificationOption) error {
+	challenge := p.hash()
+	if len(next.Challenge) != 0 && !bytes.Equal(next.Challenge, challenge) {
+		return errors.New("the challenge does not match the previous contribution's hash")
+	}
+	next.Challenge = challenge
```

Comment:
Similarly as for Phase1 imo Phase2.Verify shouldn't change the inputs. See the corresponding comment.